plugins/node/opentelemetry-instrumentation-redis-4/src/instrumentation.ts

```diff
@@ -0,0 +1,410 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  diag,
+  trace,
+  context,
+  SpanKind,
+  Span,
+  SpanStatusCode,
+} from '@opentelemetry/api';
+import {
+  isWrapped,
+  InstrumentationBase,
+  InstrumentationNodeModuleDefinition,
+  InstrumentationNodeModuleFile,
+} from '@opentelemetry/instrumentation';
+import { defaultDbStatementSerializer } from './utils';
+import { RedisInstrumentationConfig } from './types';
+import { VERSION } from './version';
+import {
+  DbSystemValues,
+  SemanticAttributes,
+} from '@opentelemetry/semantic-conventions';
+
+const OTEL_OPEN_SPANS = Symbol(
+  'opentelemetry.instruemntation.redis.open_spans'
+);
+const MULTI_COMMAND_OPTIONS = Symbol(
+  'opentelemetry.instruemntation.redis.multi_command_options'
+);
+
+interface MutliCommandInfo {
+  span: Span;
+  commandName: string;
+  commandArgs: Array<string | Buffer>;
+}
+
+const DEFAULT_CONFIG: RedisInstrumentationConfig = {
+  requireParentSpan: false,
+};
+
+export class RedisInstrumentation extends InstrumentationBase<any> {
+  static readonly COMPONENT = 'redis';
+
+  constructor(protected override _config: RedisInstrumentationConfig = {}) {
+    super('@opentelemetry/instrumentation-redis-4', VERSION, _config);
+  }
+
+  override setConfig(config: RedisInstrumentationConfig = {}) {
+    this._config = Object.assign({}, DEFAULT_CONFIG, config);
+  }
+
+  protected init() {
+    return [
+      // @node-redis/client is a new package introduced and consumed by 'redis ^4.0.0'
+      new InstrumentationNodeModuleDefinition<unknown>(
+        '@node-redis/client',
+        ['^1.0.0'],
+        (moduleExports: any, moduleVersion?: string) => {
+          diag.debug(
+            `Patching @node-redis/client@${moduleVersion} (redis@^4.x.x)`
+          );
+          return moduleExports;
+        },
+        (_moduleExports: any, moduleVersion?: string) => {
+          diag.debug(
+            `Unpatching @node-redis/client@${moduleVersion} (redis@^4.x.x)`
+          );
+        },
+        [
+          new InstrumentationNodeModuleFile<any>(
+            '@node-redis/client/dist/lib/commander.js',
+            ['^1.0.0'],
+            (moduleExports: any) => {
+              const transformCommandArguments =
+                moduleExports.transformCommandArguments;
+              if (!transformCommandArguments) {
+                this._diag.error(
+                  'internal instrumentation error, missing transformCommandArguments function'
+                );
+                return moduleExports;
+              }
+
+              // this is the function that extend a redis client with a list of commands.
+              // the function patches the commandExecutor to record a span
+              this._diag.debug('Patching redis commands executor');
+              if (isWrapped(moduleExports?.extendWithCommands)) {
+                this._unwrap(moduleExports, 'extendWithCommands');
```

Comment:
Unfortunately not :(
The package was refactored in a way that is very hard to patch and instrument. Had to do some workarounds to be able to make it work.

It could be that I missed something. If you spot a better alternative I'll be happy to discuss it.